app/src/main/java/ch/epfl/sdp/peakar/points/POICache.java

```diff
@@ -0,0 +1,155 @@
+ */
+public class POICache {
+    //Constants
+    private static final String CACHE_FILE_NAME = "POIPointsCache.txt";
+
+    private static POICache instance;
+    /*List of POIs contained in the cache*/
+    private static ArrayList<POIPoint> cachedPOIPoints;
+    /*Bounding box saved to cache*/
+    private static BoundingBox cachedBoundingBox;
+
+    /**
+     * Private constructor: the class is a singleton
+     */
+    private POICache(){
+        cachedPOIPoints = null;
+        cachedBoundingBox = null;
+    }
+
+    /**
+     * Get singleton
+     * @return POICache instance
+     */
+    public static POICache getInstance(){
+        if(instance == null)
+            instance = new POICache();
+        return instance;
+    }
+
+    /**
+     * Save downloaded POIs and bounding box to cache file
+     */
+    public void savePOIDataToCache(ArrayList<POIPoint> cachedPOIPoints, BoundingBox cachedBoundingBox, Context context){
+        //Create a new object with all needed information to save in JSON
+        POICacheContent poiCacheContent = new POICacheContent(cachedPOIPoints,cachedBoundingBox);
+        Gson gson = new Gson();
+        //Convert to JSON
+        String serializesCache = gson.toJson(poiCacheContent);
+        saveJson(serializesCache,context);
+    }
+
+    /**
+     * Save the serialized JSON to text file
+     * @param serializedObject string in json format
+     * @param context context
+     */
+    private static void saveJson(String serializedObject,Context context){
+        File cacheDir = context.getCacheDir();
+        File outputFile = new File(cacheDir,CACHE_FILE_NAME);
+        try{
+            FileWriter writer = new FileWriter(outputFile);
+            writer.append(serializedObject);
+            writer.flush();
+            writer.close();
+        }
+        catch(IOException e){
+            Log.e("Exception", "File write failed cache POIPoints: " + e.toString());
+        }
+    }
+
+
+    /**
+     * Read text file and retrieve serialized
+     * @return serialized string
+     */
+    private static String readJSON(Context context){
+        File file = new File(context.getCacheDir(),CACHE_FILE_NAME);
+        StringBuilder fileContent = new StringBuilder();
+        try{
+            BufferedReader bufferedReader = new BufferedReader(new FileReader(file));
+            String line;
+            while ((line = bufferedReader.readLine()) != null){
+                fileContent.append(line);
+                fileContent.append('\n');
+            }
+            bufferedReader.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return fileContent.toString();
+    }
+
+    /**
+     * Retrieve POI data from cache and overwrite cachedPOIPoints and cachedBoundingBox
+     * with cache file values
+     */
+    private static void retrievePOIDataFromCache(Context context){
+        String fileContent = readJSON(context);
+        Gson gson = new Gson();
```

Comment:
I am not an expert at Gson. But I see that in several function we do "new Gson()" is that something that is possible to do in the constructor once and then use the same Gson object?

Reply:
Gson is a library I used to serialize Objects directly in JSON format, I followed the lib tuto for this method

Reply:
Hmm.. Just I am not sure a new one needs to be created everytime. But as I said I am not sure...

Reply:
https://sites.google.com/site/gson/gson-user-guide

> The Gson instance does not maintain any state while invoking Json operations. So, you are free to reuse the same object for multiple Json serialization and deserialization operations.

It seems like we can indeed use only one instance!

Reply:
Ok i'll change that